### PR TITLE
Better error for invalid certificate.

### PR DESCRIPTION
**Pull request by meowjesty**
- Issue: #2824 

Improves the error logs, should be more specific now in case of an `InvalidCertificate`.

### CODE REVIEW

**Review by Razz4780 — CHANGES_REQUESTED** [review]
I'd move detecting invalid cert error to `CliError`. It offers better support for help messages/fix instructions. See `CliError::KubeAuthExecFailed` and `CliError::auth_exec_error_or`. I'd inject new logic into this function

**Comment by meowjesty** [comment]
Moved the error.

This error feels like it's nested deep into some `box`es, and I'm not sure how to reproduce the issue to get the error, that's why I'm a bit unsure on how to go about `downcast`. Do you know how to get an invalid cert error, so I can change it to `downcast` instead of checking if error `contains` a string?

File: mirrord/kube/src/error.rs [hunk 1]
```diff
@@ -7,7 +7,9 @@ pub type Result<T, E = KubeApiError> = std::result::Result<T, E>;
 
 #[derive(Debug, Error)]
 pub enum KubeApiError {
-    #[error("Kube failed: {0}")]
+    /// We manually implement `From<kube::Error>` to give a better error in case of
+    /// kube failing due to an invalid certificate.
+    #[error(transparent)]
```
Comment (Razz4780) [inline mirrord/kube/src/error.rs:12]: No longer relevant

File: mirrord/cli/src/error.rs [hunk 2]
```diff
@@ -400,7 +409,12 @@ impl CliError {
 
         match error {
             KubeApiError::KubeError(Error::Auth(AuthError::AuthExec(error))) => {
-                Self::KubeAuthExecFailed(error)
+                Self::KubeAuthExecFailed(error.to_owned())
+            }
+            KubeApiError::KubeError(Error::Service(ref fail))
+                if fail.to_string().contains("InvalidCertificate") =>
+            {
+                Self::InvalidCertificate(error)
```
Comment (Razz4780) [inline mirrord/cli/src/error.rs:417]: I'd change function name now, it's no longer just `auth_exec_error_or`. Maybe sth like `detect_known_error_or_else`? Maybe sth shorter :v

File: mirrord/cli/src/error.rs [hunk 3]
```diff
@@ -187,6 +187,15 @@ pub(crate) enum CliError {
     ))]
     AgentConnectionFailed(KubeApiError),
 
+    /// Friendlier version of the invalid certificate error that comes from a
+    /// [`kube::Error::Service`].
+    #[error("Kube API operation failed due to missing or invalid certificate: {0}")]
+    #[diagnostic(help(
+        r"1. Consider enabling `accept_invalid_certificates` in your `mirrord.json`, or;
+          2. Running `mirrord exec` with the `-c` flag."
+    ))]
```
Comment (Razz4780) [inline mirrord/cli/src/error.rs:196]:
```suggestion
    #[diagnostic("Consider enabling `accept_invalid_certificates` in your `mirrord.json` or running `mirrord exec` with the `-c` flag."))]
```

**Review by Razz4780 — CHANGES_REQUESTED** [review]

**Comment by Razz4780** [comment]
> Moved the error.
> 
> This error feels like it's nested deep into some `box`es, and I'm not sure how to reproduce the issue to get the error, that's why I'm a bit unsure on how to go about `downcast`. Do you know how to get an invalid cert error, so I can change it to `downcast` instead of checking if error `contains` a string?

I'd try something like this in the test:
1. Setup TLS server with a randomly generated self-signed certificate
2. Pass its address as `cluster_url` in [kube::Config](https://docs.rs/kube/latest/kube/struct.Config.html)
3. Create a kube client and try listing pods

File: mirrord/cli/src/error.rs [hunk 4]
```diff
@@ -462,3 +480,144 @@ impl From<OperatorApiError> for CliError {
...
+            assert!(format!("{fail:?}").contains("InvalidCertificate"),
+                "We were expecting this error {fail:?} to have `InvalidCertificate` but it's {fail}!"
+            );
+        }
+    }
+
+    /// Creates an [`hyper`] server with a broken certificate and a _catch-all_ route on
+    /// `localhost:9669`.
+    ///
+    /// The `.pem` files live in `/mirrord/tests/issue-2824-certs`.
```
Comment (Razz4780) [inline mirrord/cli/src/error.rs:564]: Can you maybe generate this cert dynamically? :eyes:

**Review by Razz4780 — APPROVED** [review]
I'd generate the test cert at runtime, but file is also fine by me

File: mirrord/cli/src/error.rs (hunk 4 above)
Comment (meowjesty) [reply mirrord/cli/src/error.rs:564]: Sure, but why? I grabbed this cert from the hyper examples repo.

Comment (Razz4780) [reply mirrord/cli/src/error.rs:564]: Dunno, just a suggestion. I don't like random files in the repo. But I don't insist ^^

Comment (meowjesty) [reply mirrord/cli/src/error.rs:575]: Changed!

**Pull request merged**